x/dsmr/partition.go

```diff
@@ -0,0 +1,217 @@
+// Copyright (C) 2024, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package dsmr
+
+import (
+	"bytes"
+	"context"
+	"encoding/binary"
+	"sort"
+
+	"github.com/ava-labs/avalanchego/ids"
+	"github.com/ava-labs/avalanchego/snow/validators"
+	"github.com/ava-labs/avalanchego/utils"
+
+	"github.com/ava-labs/hypersdk/codec"
+	"github.com/ava-labs/hypersdk/consts"
+	"github.com/ava-labs/hypersdk/internal/cache"
+)
+
+const partitionCacheSize = 100
+
+type weightedValidator struct {
+	weight uint64
+	nodeID ids.NodeID
+	// set accumulatedWeight up to index in precomputePartition
+	accumulatedWeight uint64
+}
+
+type PrecalculatedPartition[T Tx] struct {
+	validators  []*weightedValidator
+	totalWeight uint64
+}
+
+func (w *weightedValidator) Compare(o *weightedValidator) int {
+	return bytes.Compare(w.nodeID[:], o.nodeID[:])
+}
+
+func getWeightedVdrsFromState(ctx context.Context, state validators.State, pChainHeight uint64, subnetID ids.ID) ([]*weightedValidator, error) {
+	vdrs, err := state.GetValidatorSet(ctx, pChainHeight, subnetID)
+	if err != nil {
+		return nil, err
+	}
+	weightedValidators := make([]*weightedValidator, 0, len(vdrs))
+	for _, vdr := range vdrs {
+		weightedValidators = append(weightedValidators, &weightedValidator{
+			weight: vdr.Weight,
+			nodeID: vdr.NodeID,
+		})
+	}
+	return weightedValidators, nil
+}
+
+func precomputePartition[T Tx](validators []*weightedValidator) *PrecalculatedPartition[T] {
+	utils.Sort(validators)
+	accumulatedWeight := uint64(0)
+	for _, weightedVdr := range validators {
+		accumulatedWeight += weightedVdr.weight
+		weightedVdr.accumulatedWeight = accumulatedWeight
+	}
+	return &PrecalculatedPartition[T]{
+		validators:  validators,
+		totalWeight: accumulatedWeight,
+	}
+}
+
+func CalculatePartition[T Tx](ctx context.Context, state validators.State, pChainHeight uint64, subnetID ids.ID) (*PrecalculatedPartition[T], error) {
+	weightedValidators, err := getWeightedVdrsFromState(ctx, state, pChainHeight, subnetID)
+	if err != nil {
+		return nil, err
+	}
+
+	return precomputePartition[T](weightedValidators), nil
+}
+
+func calculateSponsorWeightIndex(sponsor codec.Address, totalWeight uint64) uint64 {
+	return binary.BigEndian.Uint64(sponsor[len(sponsor)-consts.Uint64Len:]) % totalWeight
+}
+
+func (pp *PrecalculatedPartition[T]) AssignTx(tx T) (ids.NodeID, bool) {
+	if pp.totalWeight == 0 || len(pp.validators) == 0 {
+		return ids.NodeID{}, false
+	}
+
+	sponsor := tx.GetSponsor()
+	sponsorWeightIndex := calculateSponsorWeightIndex(sponsor, pp.totalWeight)
+	nodeIDIndex := sort.Search(len(pp.validators), func(i int) bool {
+		return pp.validators[i].accumulatedWeight > sponsorWeightIndex
+	})
+
+	// Defensive: this should never happen
+	if nodeIDIndex >= len(pp.validators) {
+		return ids.NodeID{}, false
+	}
```

Comment:
Should we just panic and break loudly as opposed to quietly continuing to operate in a state we shouldn't be in? Wondering if leaving this check will mask a problem in the future if this does end up breaking since this case shouldn't be possible anyways...

Also `sort.Search` only returns `len(slice)` if it's not found, so we should be checking for `==` instead of `>=`.

Reply:
Removed